### PR TITLE
timegm implementation has been added to build on SunOs and AIX

### PR DESCRIPTION
**Pull request by avpalienko**
timegm function is missing on some platforms, e.g. Sun and AIX. We have to check the existence and use self-made version if not

### CODE REVIEW

File: include/private/soci-mktime.h [hunk 1]
```diff
@@ -21,6 +21,24 @@ namespace soci
 namespace details
 {
 
+namespace inner
```
Comment (vadz) [inline include/private/soci-mktime.h:24]: Do we really need this namespace? `details` is already a kind of "inner" namespace.

**Review by vadz — COMMENTED** [review]
Thanks! Is `selfmade_timegm()` your own code? A quick search finds some other versions of this function which seem a bit more clear and don't use macros, e.g. [this one](https://stackoverflow.com/a/58037981/15275).

Especially if it's your own code, it would be really nice to have at least some trivial unit tests for this function to make sure it works correctly now (which it probably does) and keeps working in the future (as it would be easy to break it with the automated tests only running under platforms not using this function at all).

File: include/private/soci-mktime.h (hunk 1 above)
Comment (avpalienko) [reply include/private/soci-mktime.h]: fixed

File: include/private/soci-mktime.h [hunk 2]
```diff
@@ -21,6 +21,24 @@ namespace soci
 namespace details
 {
 
+namespace inner 
+{
+
+time_t selfmade_timegm ( struct tm* tb );
```
[the remark this answers is not in the document]
Comment (avpalienko) [reply include/private/soci-mktime.h]: fixed

**Comment by avpalienko** [comment]
> Thanks! Is `selfmade_timegm()` your own code? A quick search finds some other versions of this function which seem a bit more clear and don't use macros, e.g. [this one](https://stackoverflow.com/a/58037981/15275).
> 
> Especially if it's your own code, it would be really nice to have at least some trivial unit tests for this function to make sure it works correctly now (which it probably does) and keeps working in the future (as it would be easy to break it with the automated tests only running under platforms not using this function at all).

Thank you for review! 

No.  It's a copy from internet. Your example does not normalize the argument but timegm does.
I've added the unit test and removed the macros

**Comment by vadz** [comment]
>  It's a copy from internet.

It would be nice to point out its provenance, if only to have a confirmation that it's available under a compatible licence.

**Comment by vadz** [comment]
Ugh, sorry, a quick code search finds this code in `mktime.c` file with

```c
/*
  Copyright (c) 1987-1997, Microsoft Corporation. All rights reserved.
*/
```

in it so we really can't include it into SOCI, unfortunately.

Could you please get the version from SO (which is implicitly in public domain AFAIK) or find some other implementation under acceptable licence?

**Comment by avpalienko** [comment]
Ups! :-(
Thank you! 
I've changed implementation to as SO

**Comment by vadz** [comment]
Merged now, thanks for the update!

**Comment by Begasus** [comment]
Thanks for this, was hitting this when checking another project that uses SOCI, fixes the build for Haiku. +1

**Pull request closed without merging**